src/soter/openssl/soter_sign_rsa.c

```diff
@@ -30,72 +30,69 @@ soter_status_t soter_sign_init_rsa_pss_pkcs8(soter_sign_ctx_t* ctx,
                                              const size_t public_key_length)
 {
     soter_status_t err = SOTER_FAIL;
-    EVP_PKEY* pkey = NULL;
     EVP_PKEY_CTX* md_pkey_ctx = NULL;
 
-    pkey = EVP_PKEY_new();
-    if (!pkey) {
-        return SOTER_NO_MEMORY;
-    }
-
-    if (!EVP_PKEY_set_type(pkey, EVP_PKEY_RSA)) {
-        goto free_pkey;
-    }
-
-    ctx->pkey_ctx = EVP_PKEY_CTX_new(pkey, NULL);
-    if (!(ctx->pkey_ctx)) {
-        err = SOTER_NO_MEMORY;
-        goto free_pkey;
+    /* soter_sign_ctx_t should be initialized only once */
+    if (!ctx || ctx->pkey || ctx->md_ctx) {
+        return SOTER_INVALID_PARAMETER;
     }
 
     if ((!private_key) && (!public_key)) {
-        err = soter_rsa_gen_key(ctx->pkey_ctx);
+        err = soter_rsa_gen_key(&ctx->pkey);
         if (err != SOTER_SUCCESS) {
-            goto free_pkey_ctx;
+            goto free_pkey;
         }
     } else {
+        ctx->pkey = EVP_PKEY_new();
+        if (!ctx->pkey) {
+            err = SOTER_NO_MEMORY;
+            goto free_pkey;
+        }
+
+        if (EVP_PKEY_set_type(ctx->pkey, EVP_PKEY_RSA) != 1) {
+            err = SOTER_FAIL;
+            goto free_pkey;
+        }
+
         if (private_key != NULL) {
-            err = soter_rsa_import_key(pkey, private_key, private_key_length);
+            err = soter_rsa_import_key(ctx->pkey, private_key, private_key_length);
             if (err != SOTER_SUCCESS) {
-                goto free_pkey_ctx;
+                goto free_pkey;
             }
         }
         if (public_key != NULL) {
-            err = soter_rsa_import_key(pkey, public_key, public_key_length);
+            err = soter_rsa_import_key(ctx->pkey, public_key, public_key_length);
             if (err != SOTER_SUCCESS) {
-                goto free_pkey_ctx;
+                goto free_pkey;
             }
         }
     }
 
     ctx->md_ctx = EVP_MD_CTX_create();
     if (!(ctx->md_ctx)) {
         err = SOTER_NO_MEMORY;
-        goto free_pkey_ctx;
+        goto free_pkey;
     }
 
     /* md_pkey_ctx is owned by ctx->md_ctx */
-    if (!EVP_DigestSignInit(ctx->md_ctx, &md_pkey_ctx, EVP_sha256(), NULL, pkey)) {
+    if (EVP_DigestSignInit(ctx->md_ctx, &md_pkey_ctx, EVP_sha256(), NULL, ctx->pkey) != 1) {
```

Comment:
> we were in safe just because use a limited set of algorithm that supported by every openssl version

Yeah, basically that. From what I've seen, the most common non-zero return code is `-2` for “algorithm not implemented”, but since the algorithms we use are included in every build of OpenSSL, I'd expect that we never had to deal with non-zero return codes in practice.

Maybe it would be more correct to handle these distinct error codes more nicely, like returning SOTER_NOT_SUPPORTED when OpenSSL says it's not supported. But keeping track of functions that do or do not return these special codes is a bother. For starters, I think we'd be definitely more safe by treating only `1` as success.